pkg/pulumiyaml/analyser.go

```diff
@@ -157,23 +417,344 @@ func (tc *typeCache) typeInvoke(ctx *evalContext, t *ast.InvokeExpr) bool {
+			Fields:              fields,
+			MaxElements:         5,
+			FieldsAreProperties: true,
+		}
+		if hint.Outputs == nil || !validReturn {
+			summary, detail := fmtr.MessageWithDetail(t.Return.Value, t.Return.Value)
+			ctx.addDiag(syntax.Error(t.Return.Syntax().Syntax().Range(), summary, detail))
+		} else {
+			tc.exprs[t] = returnType
+		}
+	} else {
+		tc.exprs[t] = hint.Outputs
+	}
+	return true
+}
+
+func (tc *typeCache) typeSymbol(ctx *evalContext, t *ast.SymbolExpr) bool {
+	var typ schema.Type = &schema.InvalidType{}
+	if root, ok := tc.resourceNames[t.Property.RootName()]; ok {
+		typ = tc.resources[root]
+	}
+	if root, ok := tc.variableNames[t.Property.RootName()]; ok {
+		typ = tc.exprs[root]
+	}
+	if root, ok := tc.configuration[t.Property.RootName()]; ok {
+		typ = root
+	}
+	runningName := t.Property.RootName()
+	setError := func(summary, detail string) *schema.InvalidType {
+		diag := syntax.Error(t.Syntax().Syntax().Range(), summary, detail)
+		ctx.addDiag(diag)
+		typ := &schema.InvalidType{
+			Diagnostics: []*hcl.Diagnostic{diag.HCL()},
+		}
+		return typ
+	}
+
+	tc.exprs[t] = typePropertyAccess(ctx, typ, runningName, t.Property.Accessors[1:], setError)
 	return true
 }
 
+func typePropertyAccess(ctx *evalContext, root schema.Type,
+	runningName string, accessors []ast.PropertyAccessor,
+	setError func(summary, detail string) *schema.InvalidType) schema.Type {
+	if len(accessors) == 0 {
+		return root
+	}
+	if root, ok := root.(*schema.UnionType); ok {
+		possibilities := map[schema.Type]struct{}{}
+		errs := []*notAssignable{}
+		for _, subtypes := range root.ElementTypes {
+			t := typePropertyAccess(ctx, subtypes, runningName, accessors,
+				func(summary, detail string) *schema.InvalidType {
+					errs = append(errs, &notAssignable{reason: summary, property: subtypes.String()})
+					return &schema.InvalidType{}
+				})
+			if _, ok := t.(*schema.InvalidType); !ok {
+				possibilities[t] = struct{}{}
+			}
+		}
+		if len(errs) > 0 {
+			op := "access"
+			if _, ok := accessors[0].(*ast.PropertySubscript); ok {
+				op = "index"
+			}
+			setError(fmt.Sprintf("Cannot %s into %s of type %s", op, runningName, root),
+				notAssignable{
+					reason:  fmt.Sprintf("'%s' could be a type that does not support %sing", runningName, op),
+					because: errs,
+				}.String())
+			return &schema.InvalidType{}
+		}
+		arr := make([]schema.Type, 0, len(possibilities))
+		for k := range possibilities {
+			arr = append(arr, k)
+		}
+		if len(arr) == 1 {
+			return arr[0]
+		}
+		return &schema.UnionType{ElementTypes: arr}
+	}
+	switch accessor := accessors[0].(type) {
+	case *ast.PropertyName:
+		properties := map[string]schema.Type{}
+		switch root := codegen.UnwrapType(root).(type) {
+		case *schema.ObjectType:
+			for _, prop := range root.Properties {
+				properties[prop.Name] = prop.Type
+			}
+		case *schema.ResourceType:
+			for _, prop := range root.Resource.Properties {
+				properties[prop.Name] = prop.Type
```

Comment:
We don't have to handle optionals here, right?

Reply:
We have to consider them, as we do on line [521](https://github.com/pulumi/pulumi-yaml/blob/iwahbe/type-checker/pkg/pulumiyaml/analyser.go#L521). I'm allowing people to assume that optional properties exist. I'm not 100% sure that I am answering your question. Let me know if I misunderstood.

Reply:
I was pattern matching against this code from the binder, where non-required props have their type wrapped in `OptionalType{}`:
 https://github.com/pulumi/pulumi/blob/36cbf572f45790f6859a8da3f8d67f6b32620fca/pkg/codegen/pcl/binder_schema.go#L273-L280

Reply:
I don't think we need to do anything like that. I'm actually very skeptical of that piece of code, since [this](https://github.com/pulumi/pulumi/blob/36cbf572f45790f6859a8da3f8d67f6b32620fca/pkg/codegen/schema/schema.go#L362-L365) is the definition of `IsRequired`. 

It means that we are doing this:
```go
		for _, prop := range src.Resource.Properties {
			typ := prop.Type
                        
			if _, optional := typ.(*schema.OptionalType); optional {
				typ = &schema.OptionalType{ElementType: typ}
			}


			properties[prop.Name] = b.schemaTypeToTypeOrConst(typ, prop)
		}
```

Reply:
Hmm, worth checking if we can simplify there, then.